homeassistant/components/ws66i/config_flow.py

```diff
@@ -0,0 +1,143 @@
+"""Config flow for WS66i 6-Zone Amplifier integration."""
+import logging
+
+from pyws66i import get_ws66i
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_IP_ADDRESS
+
+from .const import (
+    CONF_SOURCE_1,
+    CONF_SOURCE_2,
+    CONF_SOURCE_3,
+    CONF_SOURCE_4,
+    CONF_SOURCE_5,
+    CONF_SOURCE_6,
+    CONF_SOURCES,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+SOURCES = [
+    CONF_SOURCE_1,
+    CONF_SOURCE_2,
+    CONF_SOURCE_3,
+    CONF_SOURCE_4,
+    CONF_SOURCE_5,
+    CONF_SOURCE_6,
+]
+
+OPTIONS_SCHEMA = {vol.Optional(source): str for source in SOURCES}
+
+DATA_SCHEMA = vol.Schema({vol.Required(CONF_IP_ADDRESS): str})
+
+FIRST_ZONE = 11
+
+
+@core.callback
+def _sources_from_config(data):
+    sources_config = {
+        str(idx + 1): data.get(source) for idx, source in enumerate(SOURCES)
+    }
+
+    return {
+        index: name.strip()
+        for index, name in sources_config.items()
+        if (name is not None and name.strip() != "")
+    }
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    ws66i = get_ws66i(data[CONF_IP_ADDRESS])
+    await hass.async_add_executor_job(ws66i.open)
+    # No exception. run a simple test to make sure we opened correct port
+    # Test on FIRST_ZONE because this zone will always be valid
+    ret_val = await hass.async_add_executor_job(ws66i.zone_status, FIRST_ZONE)
+    if ret_val is None:
+        ws66i.close()
+        raise ConnectionError("Not a valid WS66i connection")
+
+    # Validation done. No issues. Close the connection
+    ws66i.close()
```

Comment:
Is `.close` safe to run in the event loop?

Reply:
Is a method considered safe to run on the event loop if it is not blocking? .close calls Telnet.close. The documentation there just says "close the connection". Looking for more information online, close will finish sending data on the pipeline and send EOF. It doesn't not wait for a response. To me, this sounds like it is not blocking, so safe. However the HA documentation says no I.O allowed. So is the following acceptable?
`    async def async_close(event):
        """Close the Telnet connection to the amplifier."""
        await hass.async_add_executor_job(ws66i.close)
        entry.async_on_unload(hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_close))
`
In my testing it calls close when HA shutsdown.

Reply:
Traced the code. Looks like it should be safe